File: 2400-minimum-score-after-removals-on-a-tree/minimum-score-after-removals-on-a-tree.py
class Solution:
    def minimumScore(self, nums: List[int], edges: List[List[int]]) -> int:
        N = len(nums)
        edges_map = defaultdict(list)
        for s, e in edges:
            edges_map[s].append(e)
            edges_map[e].append(s)
        
        xor_sum_tree = [0]*N
        def calculate_xor_sum(node, parent):
            current_sum = nums[node]
            for t in edges_map[node]:
                if t != parent:
                    current_sum ^= calculate_xor_sum(t, node)
            xor_sum_tree[node] = current_sum
            return current_sum
        calculate_xor_sum(0, -1)

        tree_nodes = [0] * N
        def add_tree_nodes(node, parent):
            tree_nodes[node] = set([node])
            for t in edges_map[node]:
                if t != parent:
                    tree_nodes[node] |= add_tree_nodes(t, node)
            return tree_nodes[node]
        
        add_tree_nodes(0, 0)
        # Now we consider 2 cases. Let o be tree rooted at 0, i be tree rooted at i and j be tree routed at j
        # 1. If o, i, j are not contained in eachother (assuming we romioved edge from x to i and x_dash to j).
        # 2. If j is contained in i tree. (vise versa) (before removing the corresponding edge)
        best = float('inf')
        for i in range(1, N):
            for j in range(i+1, N):
                if i in  tree_nodes[j]:
                    tree_o = xor_sum_tree[0] ^ xor_sum_tree[j]
                    tree_i = xor_sum_tree[i]
                    tree_j = xor_sum_tree[j] ^ xor_sum_tree[i] # XOR will cancell the first addition in big tree_j
                elif j in tree_nodes[i]:
                    tree_o = xor_sum_tree[0] ^ xor_sum_tree[i]
                    tree_i = xor_sum_tree[i] ^ xor_sum_tree[j] 
                    tree_j = xor_sum_tree[j]
                else:
                    tree_o = xor_sum_tree[0] ^  xor_sum_tree[i] ^ xor_sum_tree[j]
                    tree_j = xor_sum_tree[j]
                    tree_i = xor_sum_tree[i]
                best = min(best, max(tree_o, tree_i, tree_j) - min(tree_o, tree_i, tree_j))
        return best

File: 2400-minimum-score-after-removals-on-a-tree/minimum-score-after-removals-on-a-tree.go
func minimumScore(nums []int, edges [][]int) int {
    N := len(nums)

	// Create adjacency list representation of the tree
	edgesMap := make(map[int][]int)
	for _, edge := range edges {
		s, e := edge[0], edge[1]
		edgesMap[s] = append(edgesMap[s], e)
		edgesMap[e] = append(edgesMap[e], s)
	}

	// Array to store XOR sum of each subtree
	xorSumTree := make([]int, N)

	// DFS to calculate XOR sum of each subtree
	var calculateXorSum func(node, parent int) int
	calculateXorSum = func(node, parent int) int {
		currentSum := nums[node]
		for _, t := range edgesMap[node] {
			if t != parent {
				currentSum ^= calculateXorSum(t, node)
			}
		}
		xorSumTree[node] = currentSum
		return currentSum
	}
	calculateXorSum(0, -1)

	// Array to store set of nodes in each subtree
	treeNodes := make([]map[int]bool, N)
	for i := range treeNodes {
		treeNodes[i] = make(map[int]bool)
	}

	// DFS to build sets of nodes for each subtree
	var addTreeNodes func(node, parent int) map[int]bool
	addTreeNodes = func(node, parent int) map[int]bool {
		treeNodes[node][node] = true
		for _, t := range edgesMap[node] {
			if t != parent {
				childNodes := addTreeNodes(t, node)
				for child := range childNodes {
					treeNodes[node][child] = true
				}
			}
		}
		return treeNodes[node]
	}
	addTreeNodes(0, 0)

	// Initialize best score as infinity
	best := math.MaxInt32

	// Helper function to check if node x is in subtree of y
	hasNode := func(x int, yNodes map[int]bool) bool {
		return yNodes[x]
	}

	// Try all possible pairs of edges to remove
	for i := 1; i < N; i++ {
		for j := i + 1; j < N; j++ {
			var treeO, treeI, treeJ int
			// Case 1: If node i is in subtree of j
			if hasNode(i, treeNodes[j]) {
				treeO = xorSumTree[0] ^ xorSumTree[j]
				treeI = xorSumTree[i]
				treeJ = xorSumTree[j] ^ xorSumTree[i]
			// Case 2: If node j is in subtree of i
			} else if hasNode(j, treeNodes[i]) {
				treeO = xorSumTree[0] ^ xorSumTree[i]
				treeI = xorSumTree[i] ^ xorSumTree[j]
				treeJ = xorSumTree[j]
			// Case 3: If subtrees i and j are disjoint
			} else {
				treeO = xorSumTree[0] ^ xorSumTree[i] ^ xorSumTree[j]
				treeI = xorSumTree[i]
				treeJ = xorSumTree[j]
			}

			// Calculate score as max XOR - min XOR of the three components
			maxVal := max(treeO, max(treeI, treeJ))
			minVal := min(treeO, min(treeI, treeJ))
			best = min(best, maxVal-minVal)
		}
	}

	return best
}